### PR TITLE
Device plugin

### PR DESCRIPTION
**Pull request by bn222**
Supersedes https://github.com/openshift/dpu-operator/pull/26

### CODE REVIEW

File: daemon/device-plugin/types.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 Intel Corp. All Rights Reserved.
```
Comment (wizhaoredhat) [inline daemon/device-plugin/types.go:1]: Please remove this

File: daemon/device-plugin/deviceplugin.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+const (
+	VendorPluginSocketPath string = cnitypes.DaemonBaseDir + "vendor-plugin/vendor-plugin.sock"
+
+	// Device plugin settings.
+	pluginMountPath = "/var/lib/kubelet/device-plugins"
+	kubeletEndpoint = "kubelet.sock"
+	pluginEndpoint  = "sriovNet.sock"
+	resourceName    = "intel.com/ipu"
+	prefix          = "intel.com"
```
Comment (wizhaoredhat) [inline daemon/device-plugin/deviceplugin.go:29]: prefix can be removed.

File: daemon/device-plugin/deviceplugin.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+			if dev.Health != pluginapi.Healthy {
+				fmt.Printf("Error. Invalid allocation request with unhealthy device %s", id)
+			}
+
+			devName = devName + id + ","
+		}
+
+		fmt.Printf("device(s) allocated: %s\n", devName)
+		envmap := make(map[string]string)
+		envmap["NF-DEV"] = devName
```
Comment (wizhaoredhat) [inline daemon/device-plugin/deviceplugin.go:101]: Check if Eoghan's PR addresses this.

File: daemon/plugin/vendorplugin.go [hunk 4]
```diff
@@ -5,8 +5,8 @@ import (
 	"net"
 
 	"github.com/go-logr/logr"
-	pb "github.com/openshift/dpu-operator/dpu-api/gen"
 	"github.com/openshift/dpu-operator/dpu-cni/pkgs/cnitypes"
+	pb "github.com/openshift/dpu-operator/dpu-api/gen"
```
Comment (wizhaoredhat) [inline daemon/plugin/vendorplugin.go:9]: Did you run `go fmt`?

File: daemon/device-plugin/deviceplugin.go (hunk 3 above)
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go:101]: no, we are using NF-DEV now.

**Comment by wizhaoredhat** [comment]
```
go vet ./...
# github.com/openshift/dpu-operator/daemon/device-plugin
daemon/device-plugin/deviceplugin.go:120:44: cannot use &emptypb.Empty{} (value of type *emptypb.Empty) as *dpu_api.Empty value in argument to nf.client.GetDevices
```

File: daemon/device-plugin/deviceplugin.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	for _, device := range Devices.Devices {
+		nf.devices[device.ID] = pluginapi.Device{ID: device.ID, Health: pluginapi.Healthy}
+	}
+
+	for dev := range nf.devices {
+		nf.log.Info(dev)
+	}
+
+	pluginEndpoint := filepath.Join(pluginapi.DevicePluginPath, nf.socketFile)
+	fmt.Printf("Starting APF Device Plugin server at: %s\n", pluginEndpoint)
```
Comment (wizhaoredhat) [inline daemon/device-plugin/deviceplugin.go:133]: Nit: Replay APF with NF

**Pull request merged**